File: 프로그래머스/lv2/42888. 오픈채팅방/오픈채팅방.py
from collections import * 
from itertools import * 

def solution(record):
    uids = {}
    result = []
    order = []
    for input in record:
        conv = input.split()
        state = conv[0]
        if(state == "Leave"):
            uid = conv[1]
            order.append([uid,"나갔습니다."])

        elif(state == "Enter"):
            uid = conv[1]
            name = conv[2]
            order.append([uid,"들어왔습니다."])
            uids[uid] = name
        
        else:
            uid = conv[1]
            name = conv[2]
            uids[uid] = name
        
      
        
    for uid,state in order:
        result.append(uids[uid] +"님이 " + state)

    
    
    return result

File: 프로그래머스/lv2/42888. 오픈채팅방/오픈채팅방.swift
import Foundation

func solution(_ record:[String]) -> [String] {
    
    var hash: [String:String] = .init()
    var order: [(String,String)] = .init()
    var answer: [String] = .init()
    record.forEach {
        let map = $0.split{$0 == " "}.compactMap{String($0)}
        
        if map[0] == "Enter" {
            hash[map[1]] = map[2]
            order.append((map[1],"님이 들어왔습니다."))
        }
        else if map[0] == "Leave" {
             order.append((map[1],"님이 나갔습니다."))
        }
        else {
             hash[map[1]] = map[2]
        }
        
    }
    
    order.forEach{
        answer.append("\(hash[$0.0,default: "" ])\($0.1)")
    }
    
    
    return answer
}